core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -158,22 +159,29 @@ public List<Delete> getDeletesInDeleteSet() {
     return new ArrayList<>(deleteSet.values());
   }
 
-  public Optional<TransactionResult> get(Key key) throws CrudException {
+  public Optional<TransactionResult> mergeResult(Key key, Optional<TransactionResult> result)
+      throws CrudException {
     if (deleteSet.containsKey(key)) {
       return Optional.empty();
-    } else if (readSet.containsKey(key)) {
-      if (writeSet.containsKey(key)) {
-        // merge the result in the read set and the put in the write set
-        return Optional.of(
-            new TransactionResult(
-                new MergedResult(readSet.get(key), writeSet.get(key), getTableMetadata(key))));
-      } else {
-        return readSet.get(key);
-      }
+    } else if (writeSet.containsKey(key)) {
+      // merge the result in the read set and the put in the write set
+      return Optional.of(
+          new TransactionResult(
+              new MergedResult(result, writeSet.get(key), getTableMetadata(key))));
+    } else {
+      return result;
     }
-    throw new IllegalArgumentException(
-        CoreError.CONSENSUS_COMMIT_GETTING_DATA_NEITHER_IN_READ_SET_NOR_DELETE_SET_NOT_ALLOWED
-            .buildMessage());
+  }
+
+  public Optional<TransactionResult> mergeResult(
+      Key key, Optional<TransactionResult> result, Set<Conjunction> conjunctions)
+      throws CrudException {
+    return mergeResult(key, result)
+        .filter(
+            r ->
+                conjunctions.isEmpty()
+                    || !writeSet.containsKey(key)
```

Comment:
Thanks for the question. It’s to avoid unnecessary dynamic filtering. Here, we may have a merged result with our own put, or we may not. If it’s a merged result, we need to filter the result dynamically. But if it’s not a merged result, we don’t have to filter using `columnsMatchAnyOfConjunctions` with an extra cost since it’s expected to be already filtered in the underlying storage.

Reply:
Ah, got it. Thanks for the explanation.

I guess what you really want to know here is the transaction result is a merged result. If so, how about adding `isMergedResult()` or something to `TransactionResult` and using it here?

```java
    return mergeResult(key, result)
      .filter(
        r ->
          conjunctions.isEmpty()
            || !r.isMergedResult()
            || ScalarDbUtils.columnsMatchAnyOfConjunctions(r.getColumns(), conjunctions));
```

I just think it's a bit unclear that `writeSet.containsKey(key)` means the result is a merged one w/o reading `mergeResult(Key key, Optional<TransactionResult> result)`.

Reply:
That's a good idea. Thank you. Revised in 4463ef8.